### DISCUSSION TITLE
MPOLYGON support - test data needed

### DISCUSSION THREAD

**Discussion by mozman** [mozman/ezdxf#411] [Ideas]
The MPOLYGON seems to be an _AutoCAD Map_ specific entity but very similar to the HATCH entity and is has no explicit documentation in the DXF Reference. I can not create this entity in _BricsCAD_, but the MPOLYGON could be supported if I get some test data from _AutoCAD Map_ users to see if the entity is really like a HATCH entity. I would need small DXF files containing MPOLYGONS with polyline paths (LWPOLYLINE) and edge paths (ARC, ELLIPSE, SPLINE), with and without nested holes, with and without filling in many DXF versions (R2000, R2007, R2010, R2018).

A new feature will be the conversion from MPOLYGON to HATCH and the HATCH entity will get a method to extract the bounding paths as virtual as DXF primitives:
- LWPOLYLINE for a polyline path
- list of LINE, ARC, ELLIPSE and SPLINE for an edge path

If you do not want to publish the documents publicly, you can send them to ezdxf@mozman.at. The files are treated confidentially.

**Comment by mbway** [comment]
Here are some files that you might find useful. It might not be 100% coverage, I don't think there are any spline edges for example but it should be a good starting point at least.
[mpolygons.zip](https://github.com/mozman/ezdxf/files/6553652/mpolygons.zip)


Polygons were created in QGIS as ShapeFiles (.shp) and then loaded in AutoCAD 2020 using the [SpatialManager plugin](https://apps.autodesk.com/ACD/en/Detail/Index?id=5328176051015692431&appLang=en&os=Win32_64). When loading the polygons with fill (Solid) there is an issue creating gaps:
this could be the import plugin, or a rendering bug in AutoCAD. I'm not sure.
![mpolygons_with_fill_issue](https://user-images.githubusercontent.com/4923501/119819852-be571400-bee8-11eb-87c3-703aacc9d5aa.png)


![mpolygons_no_fill](https://user-images.githubusercontent.com/4923501/119819895-cc0c9980-bee8-11eb-9e17-1490acbd293d.png)
![mpolygons_with_fill](https://user-images.githubusercontent.com/4923501/119819886-c911a900-bee8-11eb-994f-206ea62921ae.png)

**Reply by mozman** [reply]
~~Please can you change the fill color, because changing the color seems to need more than just adding a group code 63. Got a "Premature end of object" error for just adding the color tag.~~

Solid fill color settings:
- ~~as ACI color index e.g. 1 for red~~,  is supported and solved for R2004+
- ~~as true color e.g. r=10, g=20, b=30 to see the color channel order~~,  is supported and solved for R2004+
- ~~as gradient~~, is supported and solved for R2004+

Nothing left to do, solid fill color for R2000 is BYLAYER.

[simple_mpolygons_with_fill_2000.zip](https://github.com/mozman/ezdxf/files/6565799/simple_mpolygons_with_fill_2000.zip)

edit: all solved for R2004+

**Comment by mozman** [comment]
Thank you! The MPOLYGON should be a HATCH variant, there is just a small difference in meaning for some DXF attributes - in theory :)

**Reply by mozman** [reply]
BricsCAD shows the MPOYGON entities without any issues:
![image](https://user-images.githubusercontent.com/4230966/119828975-054e0500-befb-11eb-84d1-8badb0a64fce.png)


BricsCAD shows a different properties palette for MPOLYGON as for HATCH, but at least the MPOLYGON is editable by grips.
